### ISSUE TITLE
JIT breaks down addmm into add and mm

### ISSUE THREAD

**Issue by fatvlady** [pytorch/pytorch#43882]
## 🐛 Bug

JIT tracing/scripting breaks down `addmm` into `add` and `mm` ops.

## To Reproduce
```
import torch
from torch.autograd import profiler
torch.set_num_threads(1)

def dense_layer(input, w, b):
    return torch.addmm(b, input, w)

if __name__ == '__main__':
    torch.random.manual_seed(1234)
    a = torch.randn(100000, 10)
    b = torch.randn(10, 10)
    c = torch.randn(10)

    with profiler.profile() as prof:
        for i in range(1000):
            dense_layer(a, b, c)
    print(prof.key_averages().table(sort_by='cpu_time_total', row_limit=3))

    traced = torch.jit.trace(dense_layer, (a, b, c))
    # traced = torch.jit.script(dense_layer)
    with profiler.profile() as prof2:
        for i in range(1000):
            traced(a, b, c)
    print(prof2.key_averages().table(sort_by='cpu_time_total', row_limit=3))
```
Output:
```
--------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  
Name            Self CPU total %  Self CPU total   CPU total %      CPU total        CPU time avg     Number of Calls  
--------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  
addmm           99.90%           5.614s           100.00%          5.620s           5.620ms          1000             
expand          0.05%            2.858ms          0.08%            4.514ms          4.514us          1000             
as_strided      0.03%            1.656ms          0.03%            1.656ms          1.656us          1000             
--------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  
Self CPU time total: 5.620s

-----------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  
Name         Self CPU total %  Self CPU total   CPU total %      CPU total        CPU time avg     Number of Calls  
-----------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  
add          65.49%           5.181s           65.74%           5.201s           5.201ms          1000             
mm           33.97%           2.687s           34.26%           2.710s           2.710ms          1000             
empty        0.28%            22.040ms         0.28%            22.040ms         11.020us         2000             
-----------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  
Self CPU time total: 7.911s
```

## Expected behavior

Both Python and traced versions should call `addmm` fused operation. Not only it is breaking correctness, it also slows down traced function.

## Environment

```
PyTorch version: 1.6.0+cpu
Is debug build: False
CUDA used to build PyTorch: None
OS: Microsoft Windows Server 2016 Datacenter
GCC version: Could not collect
Clang version: Could not collect
CMake version: version 3.16.5
Python version: 3.7 (64-bit runtime)
Is CUDA available: False
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
Versions of relevant libraries:
[pip3] msgpack-numpy==0.4.3
[pip3] numpy==1.18.1
[conda] Could not collect
```

@t-vi 

cc @gmagogsfm

**Comment by vadimkantorov** [comment]
Potentially related: https://github.com/pytorch/pytorch/issues/39661

**Comment by SplitInfinity** [comment]
Seems like this is a feature and not a bug. In the `decompose_ops` pass, we have
```
bool DecomposeOps(Block* block, CompilationUnit& decompose_funcs) {
  bool decomposed = false;
  for (auto it = block->nodes().begin(), end = block->nodes().end(); it != end;
       ++it) {
    for (auto sub : it->blocks()) {
      DecomposeOps(sub, decompose_funcs);
    }

    if (it->matches(
            "aten::addmm(Tensor self, Tensor mat1, Tensor mat2, *, Scalar beta, Scalar alpha) -> Tensor",
            /*const_inputs=*/{attr::beta, attr::alpha})) {
      // For the case where we have an addmm where alpha and beta are Attributes
      // and both of those scalars are equal to 1.0, decompose this into an mm
      // followed by an add so that it can go through the existing optimization
      // (batchmm)
```

**Comment by SplitInfinity** [comment]
It could also be the case that your program is too simple to permit fusion later on

**Comment by t-vi** [comment]
> Seems like this is a "feature" and not a "bug". In the decompose_ops pass, we have
...
> > followed by an add so that it can go through the existing optimization
> >      // (batchmm)

One of the big usecases for this is batching the matrix multiplications for the inputs of LSTMs.

If we had these fancy [e-graphs](https://egraphs-good.github.io/) we could keep both to see if batching actually applied...

**Comment by SplitInfinity** [comment]
#19988 added this optimisation but I don't understand the justification

**Comment by SplitInfinity** [comment]
#56385 is already open for fixing this.

**Comment by wanchaol** [comment]
Yeah as it says in the graph, this is a feature rather than a bug, we break down addmm to add + mm for the case that we could merge multiple mm ops together into a single batch mm, to improve the performance. This shouldn't break any correctness or slow down the most traced functions, as we merge back add and mm into addmm later if possible, see [here](https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/passes/peephole.cpp#L418). 

@fatvlady I think the slow down you hit here might because the first time static/runtime optimization that JIT is doing or some randomness, did you try a warm up before profiling? I did a try like this:
```
    # warmup
    traced(a, b, c)
    with profiler.profile() as prof2:
        for i in range(1000):
            traced(a, b, c)
    print(prof2.key_averages().table(sort_by='cpu_time_total', row_limit=3))
```

and it shows speed ups (even though on my machine when there're no warm up this is still faster)
```
--------------------  ------------  ------------  ------------  ------------  ------------  ------------  
                Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg    # of Calls  
--------------------  ------------  ------------  ------------  ------------  ------------  ------------  
         aten::addmm        62.37%        1.080s       100.00%        1.732s       1.732ms          1000  
         aten::copy_        37.33%     646.508ms        37.33%     646.508ms     646.508us          1000  
        aten::expand         0.12%       2.136ms         0.17%       3.024ms       3.024us          1000  
--------------------  ------------  ------------  ------------  ------------  ------------  ------------  
Self CPU time total: 1.732s

---------------  ------------  ------------  ------------  ------------  ------------  ------------  
           Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg    # of Calls  
---------------  ------------  ------------  ------------  ------------  ------------  ------------  
    dense_layer         0.30%       4.825ms       100.00%        1.615s       1.615ms          1000  
      aten::add        65.31%        1.055s        65.31%        1.055s       1.055ms          1000  
       aten::mm        34.31%     554.100ms        34.39%     555.410ms     555.410us          1000  
---------------  ------------  ------------  ------------  ------------  ------------  ------------  
Self CPU time total: 1.615s
```

**Comment by fatvlady** [comment]
@wanchaol I think significant slowdown can be observed on Windows platform only. On my
Linux machine both traced and non-jitted versions have similar performance, with traced being slightly faster with additional warmup iteration outside profiling region.
